test/recipes/15-test_mp_rsa.t

```diff
@@ -0,0 +1,130 @@
+
+ok(run(test(["rsa_mp_test"])), "running rsa multi prime test");
+
+my $cleartext = srctop_file("test", "rsa_mp_test_msg");
+
+my @test_param = (
+    # 3 primes, 2048-bit
+    {
+        primes => '3',
+        bits => '2048',
+    },
+    # 4 primes, 2048-bit
+    {
+        primes => '4',
+        bits => '2048',
+    },
+    # 8 primes, 2048-bit
+    {
+        primes => '8',
+        bits => '2048',
+    },
+    # 15 primes, 2048-bit
+    {
+        primes => '15',
+        bits => '2048',
+    },
+    # 8 primes, 15360-bit (3 & 4 primes for 15360 bit is too long to gen a key)
+    {
+        primes => '8',
+        bits => '15360',
+    },
+    # 15 primes, 15360-bit
+    {
+        primes => '15',
+        bits => '15360',
+    },
+);
+
+# genrsa
+run_mp_tests(0);
+# evp
+run_mp_tests(1);
+
+sub run_mp_tests {
+    my $evp = shift;
+
+    foreach my $param (@test_param) {
+        my $name = "evp$param->{bits}p$param->{primes}";
+        my $primes = $param->{primes};
+        my $bits = $param->{bits};
+
+        if (!$evp) {
+            $name = "$param->{bits}p$param->{primes}";
+        }
+
+        if ($evp) {
+            ok(run(app([ 'openssl', 'genpkey', '-out', 'rsamptest.pem',
+                         '-algorithm', 'RSA', '-pkeyopt', "rsa_keygen_primes:$primes",
+                         '-pkeyopt', "rsa_keygen_bits:$bits"])), "genrsa $name");
+        } else {
+            ok(run(app([ 'openssl', 'genrsa', '-out', 'rsamptest.pem',
+                         '-primes', $primes, $bits])), "genrsa $name");
+        }
+
+        ok(run(app([ 'openssl', 'rsa', '-check', '-in', 'rsamptest.pem',
+                     '-noout'])), "rsa -check $name");
+        if ($evp) {
+            ok(run(app([ 'openssl', 'pkeyutl', '-inkey', 'rsamptest.pem',
+                         '-encrypt', '-in', "$cleartext",
+                         '-out', 'rsamptest.enc' ])), "rsa $name encrypt");
+            ok(run(app([ 'openssl', 'pkeyutl', '-inkey', 'rsamptest.pem',
+                         '-decrypt', '-in', 'rsamptest.enc',
+                         '-out', 'rsamptest.dec' ])), "rsa $name decrypt");
+        } else {
+            ok(run(app([ 'openssl', 'rsautl', '-inkey', 'rsamptest.pem',
+                         '-encrypt', '-in', "$cleartext",
+                         '-out', 'rsamptest.enc' ])), "rsa $name encrypt");
+            ok(run(app([ 'openssl', 'rsautl', '-inkey', 'rsamptest.pem',
+                         '-decrypt', '-in', 'rsamptest.enc',
+                         '-out', 'rsamptest.dec' ])), "rsa $name decrypt");
+        }
+
+        ok(check_msg(), "rsa $name check result");
+
+        # clean up temp files
+        unlink 'rsamptest.pem';
+        unlink 'rsamptest.enc';
+        unlink 'rsamptest.dec';
+    }
+}
+
+sub check_msg {
+    open(my $fh, "<", "$cleartext") or return 0;
+    my @msg = <$fh>;
+    close $fh;
+    open($fh, "<", "rsamptest.dec") or return 0;
+    my @dec = <$fh>;
+    close $fh;
+
```

Comment:
Just to make absolutely sure not to end up getting false negatives because of differing line endings, you should add this here:

``` perl
    # Avoid false negatives because of differing line endings
    @msg = map { s|\R$||; $_ } (@msg);
    @dec = map { s|\R$||; $_ } (@dec);

```

Reply:
What does that `\R` in the regex mean?

Reply:
Ahh, it matches the new line character and was added in perl v5.10...

Reply:
> Just to make absolutely sure not to end up getting false negatives.

But this makes you formally unsure of the outcome. I mean what if `message\n` is decrypted as `message\r`? BTW, if test message is a temporary file, i.e. one generated by this script, it should eliminate ambiguity.

Reply:
`\R` will strip 'generic' newline character(s) so seems no problem with `s|\R$||`. And it's more convenient to maintain the plain text outside code by using a non-temporary file...

Reply:
Correctness first, convenience second. Once again, if you compare effectively truncated messages, and that's what happens, `s|\R$||` truncates messages, how can you be *sure*, not "pretty much sure" but really *sure*, that if was processed correctly?

Reply:
So how about this: reading plain text from a file -> strip the newline chars -> write the plain text to a temp file -> do the real test -> check against the decrypted text

Thus both correctness and convenience are gained...

Reply:
I just had a look at `apps/pkeyutl.c` and `apps/rsautl.c`, and confirmed for myself that they treat the input and output files as binary, so line endings shouldn't be a factor here.  As a matter of fact, we should treat the files as binary here as well then...  So the best way to compare them might be this:

``` perl
sub check_msg {
    my $msg;
    my $dec;

    open(my $fh, "<", "$cleartext") or return 0;
    binmode $fh;
    read($fh, $msg, 10240);    # We know the cleartext is smaller than 10KiB
    close $fh;
    open($fh, "<", "rsamptest.dec") or return 0;
    binmode $fh;
    read($fh, $dec, 10240);    # We know the cleartext is smaller than 10KiB
    close $fh;

    if ($msg ne $dec) {
        print STDERR "cleartext and decrypted are not the same";
        return 0;
    }
    return 1;
}
```

Reply:
> we should treat the files as binary here as well then...

I agree with this...